Assets/LeapMotionModules/Interaction/Scripts/InteractionController.cs

```diff
@@ -0,0 +1,1382 @@
+    /// ReleaseGrasp() was manually called on the controller.
+    /// </summary>
+    protected virtual void onGraspedObjectForciblyReleased(IInteractionBehaviour objectToBeReleased) { }
+
+    /// <summary>
+    /// Returns whether this controller should grasp an object this fixed frame, and if so,
+    /// sets objectToGrasp to the object the controller should grasp.
+    /// </summary>
+    protected abstract bool checkShouldGrasp(out IInteractionBehaviour objectToGrasp);
+
+    /// <summary>
+    /// Returns whether this controller should release an object this fixed frame, and if so,
+    /// sets objectToRelease to the object the controller should release.
+    /// </summary>
+    protected abstract bool checkShouldRelease(out IInteractionBehaviour objectToRelease);
+
+    private List<InteractionController> _releasingControllersBuffer = new List<InteractionController>();
+    /// <summary>
+    /// Releases the object this hand is holding and returns true if the hand was holding an object,
+    /// or false if there was no object to release. The released object will dispatch OnGraspEnd()
+    /// immediately. The hand is guaranteed not to be holding an object directly after this method
+    /// is called.
+    /// </summary>
+    public bool ReleaseGrasp() {
+      if (_graspedObject == null) {
+        return false;
+      }
+      else {
+        _releasingControllersBuffer.Clear();
+        _releasingControllersBuffer.Add(this);
+
+        onGraspedObjectForciblyReleased(_graspedObject);
+
+        _graspedObject.EndGrasp(_releasingControllersBuffer);
+        _graspedObject = null;
+
+        return true;
+      }
+    }
+
+    /// <summary>
+    /// As ReleaseGrasp(), but also outputs the released object into releasedObject if the hand
+    /// successfully released an object.
+    /// </summary>
+    public bool ReleaseGrasp(out IInteractionBehaviour releasedObject) {
+      releasedObject = _graspedObject;
+
+      if (ReleaseGrasp()) {
+        // releasedObject will be non-null
+        return true;
+      }
+
+      // releasedObject will be null
+      return false;
+    }
+
+    /// <summary>
+    /// Attempts to release this hand's object, but only if the argument object is the object currently
+    /// grasped by this hand. If the hand was holding the argument object, returns true, otherwise returns false.
+    /// </summary>
+    public bool ReleaseObject(IInteractionBehaviour toRelease) {
+      if (_graspedObject == toRelease) {
+        ReleaseGrasp();
+        return true;
+      }
+      else {
+        return false;
+      }
+    }
+
+    #region Grasp State Checking
+
+    /// <summary>
+    /// Called by the Interaction Manager every fixed frame.
+    /// Returns true if the hand just released an object and outputs the released object into releasedObject.
+    /// </summary>
+    public bool CheckGraspEnd(out IInteractionBehaviour releasedObject) {
```

Comment:
Ahhh yeah that makes perfect sense.  Maybe we could do something similar to what happened with the editor functions in the graphic renderer, make an additional class for 'internal' functions.  So the interaction manager would do things like:

```
controller.internal.CheckEndingGrasp
```

Since it seems that the user should definitely never be calling these methods themselves right?